modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/SortedExternalResult.java

```diff
@@ -0,0 +1,475 @@
+
+        lastWrittenPos = initFilePos + written;
+
+        chunks.add(new Chunk(initFilePos, lastWrittenPos));
+
+        for (Value[] row : rows) {
+            long delta = H2Utils.calculateMemoryDelta(null, row, null);
+
+            memTracker.released(-delta);
+        }
+    }
+
+    /**
+     * Adds row to hash index.
+     *
+     * @param row Row.
+     * @param rowPosInFile Row position in file.
+     */
+    private void addRowToHashIndex(Value[] row, long rowPosInFile) {
+        ValueRow distKey = getRowKey(row);
+
+        hashIdx.put(distKey, rowPosInFile);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void reset() {
+        spillRowsBufferToDisk();
+
+        if (resQueue != null) {
+            resQueue.clear();
+
+            for (Chunk chunk : chunks)
+                chunk.reset();
+        }
+        else {
+            resQueue = sort == null ? new LinkedList<>() : new PriorityQueue<>(new Comparator<Chunk>() {
+                @Override public int compare(Chunk o1, Chunk o2) {
+                    int c = sort.compare(o1.currentRow(), o2.currentRow());
+
+                    if (c != 0)
+                        return c;
+
+                    // Compare batches to ensure they emit rows in the arriving order.
+                    return Long.compare(o1.start, o2.start);
+                }
+            });
+        }
+
+        // Init chunks.
+        for (Chunk chunk : chunks) {
+            if (chunk.next())
+                resQueue.offer(chunk);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Value[] createEmptyArray(int colCnt) {
+        return new Value[colCnt];
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void onClose() {
+        super.onClose();
+
+        U.closeQuiet(hashIdx);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean contains(Value[] values) {
+        return getPreviousRow(values) != null;
+    }
+
+    /** {@inheritDoc} */
+    @Override public synchronized ResultExternal createShallowCopy() {
+        onChildCreated();
+
+        return new SortedExternalResult(this);
+    }
+
+    /**
+     * @return whether this result is a distinct result
+     */
+    private boolean isAnyDistinct() {
+        return distinct || distinctIndexes != null;
+    }
+
+    /**
+     * Extracts distinct row key from the row.
+     * @param row Row.
+     * @return Distinct key.
+     */
+    public ValueRow getRowKey(Value[] row) {
+        if (distinctIndexes != null) {
+            int cnt = distinctIndexes.length;
+
+            Value[] newValues = new Value[cnt];
+
+            for (int i = 0; i < cnt; i++)
+                newValues[i] = row[distinctIndexes[i]];
+
+            row = newValues;
+        } else if (row.length > visibleColCnt)
+            row = Arrays.copyOf(row, visibleColCnt);
+
+        return ValueRow.get(row);
+    }
+
+    /**
+     * Sorted rows chunk on the disk.
+     */
+    private class Chunk {
```

Comment:
Let's make class static

Reply:
It uses `readRowFromFile()` from parent.